cdap-app-fabric/src/main/java/co/cask/cdap/metadata/FieldLineageAdmin.java

```diff
@@ -0,0 +1,308 @@
+import co.cask.cdap.proto.metadata.lineage.ProgramInfo;
+import co.cask.cdap.proto.metadata.lineage.ProgramRunOperations;
+import com.google.inject.Inject;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+
+/**
+ * Service to compute field lineage based on operations stored in {@link DefaultFieldLineageReader}.
+ */
+public class FieldLineageAdmin {
+
+  private final FieldLineageReader fieldLineageReader;
+
+  @Inject
+  FieldLineageAdmin(FieldLineageReader fieldLineageReader) {
+    this.fieldLineageReader = fieldLineageReader;
+  }
+
+  /**
+   * Get the set of fields written to the EndPoint by field lineage {@link WriteOperation},
+   * over the given time range, optionally prefixed with the given prefix.
+   *
+   * @param endPoint the EndPoint for which the fields need to be returned
+   * @param start start time (inclusive) in milliseconds
+   * @param end end time (exclusive) in milliseconds
+   * @param prefix prefix for the field name, if {@code null} then all fields are returned
+   * @return set of fields written to a given EndPoint
+   */
+  public Set<String> getFields(EndPoint endPoint, long start, long end, @Nullable String prefix) {
+    Set<String> fields = fieldLineageReader.getFields(endPoint, start, end);
+    if (prefix == null) {
+      return fields;
+    }
+
+    Set<String> prefixFields = new HashSet<>();
+    for (String field : fields) {
+      if (field.startsWith(prefix)) {
+        prefixFields.add(field);
+      }
+    }
+    return prefixFields;
+  }
+
+  /**
+   * Get the summary for the specified EndPointField over a given time range depending on the direction specified.
+   * Summary in the "incoming" direction consists of set of EndPointFields which participated in the computation
+   * of the given EndPointField; while summary in the "outgoing" direction consists of set of EndPointFields
+   * which were computed from the specified EndPointField. When direction is specified as 'both', incoming as well
+   * as outgoing summaries are returned.
+   *
+   * @param direction the direction can one of the "incoming", "outgoing", or "both"
+   * @param endPointField the EndPointField for which incoming summary to be returned
+   * @param start start time (inclusive) in milliseconds
+   * @param end end time (exclusive) in milliseconds
+   * @return the FieldLineageSummary
+   */
+  FieldLineageSummary getSummary(String direction, EndPointField endPointField, long start, long end) {
+    Set<DatasetField> incoming = null;
+    Set<DatasetField> outgoing = null;
+    if (direction.equals(Constants.FieldLineage.Direction.INCOMING)
+            || direction.equals(Constants.FieldLineage.Direction.BOTH)) {
```

Comment:
indentation looks strange. here and several other places below

Reply:
fixed.